### PR TITLE
[BugFix][Branch-2.2] Fix invalid ref to counter in `HdfsParquetScanner`

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [x] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

I just came across a buggy case, and I think it could be general cases, so I bring it out and hope it would be helpful to you.
In our object pool implementation, we release object in reversed order.
- If we allocate object in A, B, C
- then we release object in C, B, A
- And we have to make sure that, in dtor of B, we don't use C. 

----- 
But that case happens to me.   We have `HiveDataSource`, which has a object pool `pool` it allocates HdfsParquetScanner first. and in HdfsParquetScanner, it allocates HdfsParquetProfile

```
HiveDataSource (pool)
 -> pool.add(new HdfsParquetScanner()) -> scanner
   -> pool.add(new HdfsParquetProfile()) -> profile
~HdfsParquetScanner() {
  close()
 }
HdfsParquetScanner::close() {
  update_parquet_profile(profile)
}
```

In normal case, everything works fine. But when a user cancels a query(to me it's because of memory limit), it will call ~HiveDataSource, and pool will release objects in following order: 
- release profile
- release scanner (calling ~HdfsParquetScanner)

In ~HdfsParquetScanner, we try to update profile. However right now `profile` object has been released, 

----
So to summarize the problem is that:
- A has a object pool 
- A allocates two objects B, C.  And B holds C
- when ~A is called, C will be released first, and B will be released then.
- But in ~B we might use C. The bug happens.

### CODE REVIEW

**Comment by imay** [comment]
Is this bug in main branch?

**Review by imay — APPROVED** [review]

**Review by DorianZheng — APPROVED** [review]

**Comment by dirtysalt** [comment]
> Is this bug in main branch?

Yes, Fixed with this PR https://github.com/StarRocks/starrocks/pull/7478
